pydicom/valuerep.py

```diff
@@ -43,6 +43,318 @@
+)
+IS_REGEX = re.compile(r"^ *[+\-]?\d+ *$")
+DA_REGEX = re.compile(_range_regex(r"\d{4}(0[1-9]|1[0-2])([0-2]\d|3[01])"))
+DT_REGEX = re.compile(_range_regex(
+    r"\d{4}((0[1-9]|1[0-2])(([0-2]\d|3[01])(([01]\d|2[0-3])"
+    r"([0-5]\d((60|[0-5]\d)(\.\d{1,6} ?)?)?)?)?)?)?([+-][01]\d\d\d)?")
+)
+TM_REGEX = re.compile(_range_regex(
+    r"([01]\d|2[0-3])([0-5]\d((60|[0-5]\d)(\.\d{1,6} ?)?)?)?"))
+UR_REGEX = re.compile(r"^[A-Za-z_\d:/?#\[\]@!$&'()*+,;=%\-.~]* *$")
+
+
+def validate_vr_length(vr: str, value: Union[str, bytes]) -> Tuple[bool, str]:
+    """Validate the value length for a given VR.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str or bytes
+        The value to validate.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    max_length = MAX_VALUE_LEN.get(vr, 0)
+    if max_length > 0:
+        value_length = len(value)
+        if value_length > max_length:
+            return False, (
+                f"The value length ({value_length}) exceeds the "
+                f"maximum length of {max_length} allowed for VR {vr}."
+            )
+    return True, ""
+
+
+def validate_regex(vr: str, value: Union[str, bytes],
+                   regex: Any) -> Tuple[bool, str]:
+    """Validate the value for a given VR for allowed characters
+    using a regular expression.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str
+        The value to validate.
+    regex : re.Pattern
+        The precompiled regex pattern used for matching.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    if not re.match(regex, value):
+        return False, f"Invalid value for VR {vr}: '{value!r}'"
+    return True, ""
+
+
+def validate_length_and_regex(vr: str, value: Union[str, bytes],
+                              regex: Any) -> Tuple[bool, str]:
+    """Validate the value for a given VR both for maximum length and
+    for allowed characters using a regular expression.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str
+        The value to validate.
+    regex : re.Pattern
+        The precompiled regex pattern used for matching.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    is_valid_len, msg1 = validate_vr_length(vr, value)
+    is_valid_expr, msg2 = validate_regex(vr, value, regex)
+    msg = " ".join([msg1, msg2]).strip()
+    if msg:
+        msg += (
+            " Please see <https://dicom.nema.org/medical/dicom/current/output"
+            "/html/part05.html#table_6.2-1> for allowed values for each VR."
+        )
+    return is_valid_len and is_valid_expr, msg
+
+
+def validate_ae(vr: str, value: Union[str, bytes]) -> Tuple[bool, str]:
```

Comment:
Allows "\n" in `value`

Reply:
Hm, turns out that a single newline at the end of a string is ignored by the regex matching - I wasn't aware of this, and didn't find documentation about this yet, especially about how to change this behavior. In the worst case, I can check this separately.

Reply:
From the doc:
> $   Matches the end of the string or just before the newline at the end of the string

This sounds like the problem...

Reply:
Couldn't get it to work with regex alone, so I added an extra check. Maybe somebody has a better idea...